src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/config/RedisConfiguration.kt

```diff
@@ -1,16 +1,138 @@
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.client.ClientResult
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.model.community.StaffMember
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.model.shouldNotBeReached
+import java.time.Duration
 
 @Configuration
+@EnableCaching
 class RedisConfiguration {
   @Bean
-  fun redisTemplate(connectionFactory: RedisConnectionFactory): RedisTemplate<String, String> {
-    val template = RedisTemplate<String, String>()
-    template.connectionFactory = connectionFactory
-    return template
+  fun redisCacheManagerBuilderCustomizer(
+    buildProperties: BuildProperties,
+    objectMapper: ObjectMapper
+  ): RedisCacheManagerBuilderCustomizer? {
+    val version = buildProperties.version
+
+    return RedisCacheManagerBuilderCustomizer { builder: RedisCacheManagerBuilder ->
+      builder.clientCacheFor<List<StaffMember>>("staffMembersCache", Duration.ofHours(6), version, objectMapper)
+        .clientCacheFor<StaffMember>("staffMemberCache", Duration.ofHours(6), version, objectMapper)
+    }
+  }
+
+  private inline fun <reified T> RedisCacheManagerBuilder.clientCacheFor(cacheName: String, duration: Duration, version: String, objectMapper: ObjectMapper) =
+    this.withCacheConfiguration(
+      cacheName,
+      RedisCacheConfiguration.defaultCacheConfig()
+        .entryTtl(duration)
+        .serializeValuesWith(SerializationPair.fromSerializer(ClientResultRedisSerializer(objectMapper, object : TypeReference<T>() {})))
+        .prefixCacheNameWith(version)
+    )
+}
+
+class ClientResultRedisSerializer(
+  private val objectMapper: ObjectMapper,
+  private val typeReference: TypeReference<*>
+) : RedisSerializer<ClientResult<Any>> {
+  override fun serialize(clientResult: ClientResult<Any>?): ByteArray? {
+    val toSerialize = when (clientResult) {
+      is ClientResult.StatusCodeFailure -> {
+        SerializableClientResult(
+          discriminator = ClientResultDiscriminator.STATUS_CODE_FAILURE,
+          status = clientResult.status,
+          body = clientResult.body,
+          exceptionMessage = null,
+          type = null,
+          method = clientResult.method,
+          path = clientResult.path
+        )
+      }
+      is ClientResult.OtherFailure -> {
+        SerializableClientResult(
+          discriminator = ClientResultDiscriminator.OTHER_FAILURE,
+          status = null,
+          body = null,
+          exceptionMessage = clientResult.exception.message,
+          type = null,
+          method = clientResult.method,
+          path = clientResult.path
+        )
+      }
+      is ClientResult.Success -> {
+        SerializableClientResult(
+          discriminator = ClientResultDiscriminator.SUCCESS,
+          status = clientResult.status,
+          body = objectMapper.writeValueAsString(clientResult.body),
+          exceptionMessage = null,
+          type = clientResult.body::class.java.typeName,
+          method = null,
```

Comment:
are these `discriminator` values the key to whether the call to the cache is a hit or a miss?

Ought this logic to be described in a unit test?

Reply:
These identify the type of ClientResult that's being cached:
 - Success (HTTP response was 2xx response)
 - StatusCodeFailure (HTTP response was 4xx, 5xx)
 - OtherFailure (Broken pipe, Unknown host etc.)

This allows us to construct the correct implementation of ClientResult fom what's stored in Redis when the cache is hit.  Doing this lets us continue to write code as though the cache isn't even there - it appears to the Client call-sites that the request has been made.

The actual logic of hit/miss/TTL etc. is handled by Spring Cache so nothing for us to do there.

Reply:
I have added some unit tests around this though